clouddrift/adapters/gdp.py

```diff
@@ -0,0 +1,548 @@
+import numpy as np
+import pandas as pd
+from datetime import datetime
+import xarray as xr
+import urllib.request
+import concurrent.futures
+import re
+import tempfile
+from tqdm import tqdm
+import os
+import warnings
+
+GDP_VERSION = "2.00"
+GDP_DATA_URL = "https://www.aoml.noaa.gov/ftp/pub/phod/lumpkin/hourly/v2.00/netcdf/"
+GDP_FILENAME_PATTERN = "drifter_{id}.nc"
+GDP_TMP_PATH = os.path.join(tempfile.gettempdir(), "clouddrift", "gdp")
+
+
+def parse_directory_file(filename: str) -> pd.DataFrame:
+    """Read a directory file which contains metadata of drifters' releases.
+
+    Due to naming of these files, a manual update of the last file name after an
+    update of the dataset is needed.
+
+    Args:
+        filename (str): Name of the directory file
+    Returns:
+        pd.DataFrame: Sorted list of drifters
+    """
+
+    GDP_DIRECTORY_FILE_URL = "https://www.aoml.noaa.gov/ftp/pub/phod/buoydata/"
+    df = pd.read_csv(
+        os.path.join(GDP_DIRECTORY_FILE_URL, filename), delimiter="\s+", header=None
+    )
+
+    # Combine the date and time columns to easily parse dates below.
+    df[4] += " " + df[5]
+    df[8] += " " + df[9]
+    df[12] += " " + df[13]
+    df = df.drop(columns=[5, 9, 13])
+    df.columns = [
+        "ID",
+        "WMO_number",
+        "program_number",
+        "buoys_type",
+        "Deployment_date",
+        "Deployment_lat",
+        "Deployment_lon",
+        "End_date",
+        "End_lat",
+        "End_lon",
+        "Drogue_off_date",
+        "death_code",
+    ]
+    for t in ["Deployment_date", "End_date", "Drogue_off_date"]:
+        df[t] = pd.to_datetime(df[t], format="%Y/%m/%d %H:%M", errors="coerce")
+
+    return df
+
+
+def get_gdp_metadata() -> pd.DataFrame:
+    """Download and parse GDP metadata and return it as a Pandas DataFrame."""
+
+    directory_file_names = [
+        "dirfl_1_5000.dat",
+        "dirfl_5001_10000.dat",
+        "dirfl_10001_15000.dat",
+        "dirfl_15001_jul22.dat",
+    ]
+    df = pd.concat([parse_directory_file(f) for f in directory_file_names])
+    df.sort_values(["Deployment_date"], inplace=True, ignore_index=True)
+    return df
+
+
+def order_by_date(df: pd.DataFrame, idx: list[int]) -> np.ndarray[int]:
+    """From the previously sorted directory files DataFrame, return the drifter
+    indices sorted by their end date.
+
+    Args:
+        idx [list]: List of drifters to include in the ragged array
+    Returns:
+        idx [list]: Sorted list of drifters
+    """
+    return df.ID[np.where(np.in1d(df.ID, idx))[0]].values
+
+
+def fetch_netcdf(url: str, file: str):
+    """Download and save the file from the given url, if not already downloaded."""
+    if not os.path.isfile(file):
```

Comment:
This only looks if the file exists in the directory. Ideally, that would also need to compare if the server's version is newer.

So locally, you can get the file creation/modification with `os.path`:
```
# file creation
c_timestamp = os.path.getctime(path)

# file modification
timestamp = os.path.getmtime(path)
```
but I'm not sure if there is a way to do this with `urllib`.

Reply:
Yes, if we make a HEAD request, one of the headers returns the last modified time. Something like

```python
req = urllib.request.Request(url, method='HEAD')
resp = urllib.request.urlopen(req)
last_modified_string = resp.headers.get('Last-modified')
```

(didn't test)

The same HEAD request will also return the `Content-length` header which can be used to compare with the file size on disk and determine if the file was interrupted.

This will significantly slow down the download because we're adding an additional request per file (the files are small so the batch download is latency-bound rather than bandwidth-bound). So if we do it, I'd make it optional with an argument.

Reply:
Great. We may as well just switch completely to `requests` instead of `urllib` as it's already a dependency and has an overall simpler API.

Reply:
Agree

Reply:
FYI: I did a small test. For ten files, it takes ~3.5s to retrieve the headers and compare the times. And it takes ~11s to download.

Reply:
That's pretty good, IMO, worth it having the check on. In favor of keeping it simple, I change my mind and suggest just always doing this until somebody complains. We can add an option to disable the check then.

Reply:
Opened #95 to track this.